homeassistant/components/zha/api.py

```diff
@@ -0,0 +1,419 @@
+                cluster_type=cluster_type,
+                manufacturer=manufacturer
+            )
+        _LOGGER.debug("Set attribute for: %s %s %s %s %s %s %s",
+                      "{}: [{}]".format(ATTR_CLUSTER_ID, cluster_id),
+                      "{}: [{}]".format(ATTR_CLUSTER_TYPE, cluster_type),
+                      "{}: [{}]".format(ATTR_ENTITY_ID, entity_id),
+                      "{}: [{}]".format(ATTR_ATTRIBUTE, attribute),
+                      "{}: [{}]".format(ATTR_VALUE, value),
+                      "{}: [{}]".format(ATTR_MANUFACTURER, manufacturer),
+                      "{}: [{}]".format(RESPONSE, response)
+                      )
+
+    hass.services.async_register(DOMAIN, SERVICE_SET_ZIGBEE_CLUSTER_ATTRIBUTE,
+                                 set_zigbee_cluster_attributes,
+                                 schema=SERVICE_SCHEMAS[
+                                     SERVICE_SET_ZIGBEE_CLUSTER_ATTRIBUTE
+                                 ])
+
+    async def issue_zigbee_cluster_command(service):
+        """Issue command on zigbee cluster on zha entity."""
+        entity_id = service.data.get(ATTR_ENTITY_ID)
+        cluster_id = service.data.get(ATTR_CLUSTER_ID)
+        cluster_type = service.data.get(ATTR_CLUSTER_TYPE)
+        command = service.data.get(ATTR_COMMAND)
+        command_type = service.data.get(ATTR_COMMAND_TYPE)
+        args = service.data.get(ATTR_ARGS)
+        manufacturer = await _get_manufacturer(service.data)
+        component = hass.data.get(entity_id.split('.')[0])
+        entity = component.get_entity(entity_id)
+        response = None
+        if entity is not None:
+            response = await entity.issue_cluster_command(
+                cluster_id,
+                command,
+                command_type,
+                args,
+                cluster_type=cluster_type,
+                manufacturer=manufacturer
+            )
+        _LOGGER.debug("Issue command for: %s %s %s %s %s %s %s %s",
+                      "{}: [{}]".format(ATTR_CLUSTER_ID, cluster_id),
+                      "{}: [{}]".format(ATTR_CLUSTER_TYPE, cluster_type),
+                      "{}: [{}]".format(ATTR_ENTITY_ID, entity_id),
+                      "{}: [{}]".format(ATTR_COMMAND, command),
+                      "{}: [{}]".format(ATTR_COMMAND_TYPE, command_type),
+                      "{}: [{}]".format(ATTR_ARGS, args),
+                      "{}: [{}]".format(ATTR_MANUFACTURER, manufacturer),
+                      "{}: [{}]".format(RESPONSE, response)
+                      )
+
+    hass.services.async_register(DOMAIN, SERVICE_ISSUE_ZIGBEE_CLUSTER_COMMAND,
+                                 issue_zigbee_cluster_command,
+                                 schema=SERVICE_SCHEMAS[
+                                     SERVICE_ISSUE_ZIGBEE_CLUSTER_COMMAND
+                                 ])
+
+    @websocket_api.async_response
+    async def websocket_entities_by_ieee(hass, connection, msg):
```

Comment:
Why define this inside the function and not outside ?

Reply:
I just mirrored how it was for the services in the init module before. I can split them all out of the function.

Reply:
this one actually captures the listener object from the function definition. I can move it and store a reference to the listener in hass.data if that is preferred? I moved the ones that don't capture listener to outside of the function for now.